from controller import Robot
import random

# ================== ุงูุฅุนุฏุงุฏ (Setup) ==================
robot = Robot()
timestep = int(robot.getBasicTimeStep())

# 1. ุฃุฌูุฒุฉ ุงูุฑุคูุฉ
cam_reg = robot.getDevice("cam_reg")
if cam_reg:
    cam_reg.enable(timestep)
    cam_reg.recognitionEnable(timestep)



# ุฅุถุงูุฉ ุงููุงููุฑุง ุงูุนูููุฉ
top_cam = robot.getDevice("partner_cam") # ุงูุงุณู ุงูุฐู ุงุฎุชุฑุชู
if top_cam:
    top_cam.enable(timestep)
    top_cam.recognitionEnable(timestep)
else:
    print("โ๏ธ ุชุญุฐูุฑ: ูู ูุชู ุงูุนุซูุฑ ุนูู ุงููุงููุฑุง ุงูุนูููุฉ ุงููุณูู 'Camera'")
# 2. ุฃุฌูุฒุฉ ุงูุญุฑูุฉ
left_motor = robot.getDevice("left wheel")
right_motor = robot.getDevice("right wheel")
left_motor.setPosition(float('inf'))
right_motor.setPosition(float('inf'))
left_motor.setAvailableTorque(10.0) # ุฒูุงุฏุฉ ุงูุนุฒู ุงููุชุงุญ
right_motor.setAvailableTorque(10.0)
left_motor.setVelocity(0.0)
right_motor.setVelocity(0.0)



HANDOVER_STOP_DIST = 0.30  # ูุชุฑ (ุชูุถุจุท ุจุงูุชุฌุฑุจุฉ)
HANDOVER_ARM_DELAY = 25   # ุนุฏุฏ ุฎุทูุงุช ูุจู ุฅูุฒุงู ุงูุฐุฑุงุน

# 3. ุงูุฐุฑุงุน ูุงููููุท
wrist_roll = robot.getDevice("wrist_roll")
delivered_cubes = 0

arm_pitch = robot.getDevice("arm_pitch")
gripper_left = robot.getDevice("gripper_left")
gripper_right = robot.getDevice("gripper_right")

# 4. ูุณุชุดุนุฑุงุช ุงููููุท
gl_sensor = robot.getDevice("gripper_left_sensor")
gr_sensor = robot.getDevice("gripper_right_sensor")
if gl_sensor: gl_sensor.enable(timestep)
if gr_sensor: gr_sensor.enable(timestep)

# 5. ุญุณุงุณุงุช ุงูุณููุงุฑ (ููุญูุงูุฉ ูู ุงูุฑูุจูุชุงุช ุงูุฃุฎุฑู)# 5. ุญุณุงุณุงุช ุงูุณููุงุฑ - ูุญุงููุฉ ุชุนุฑูู ูุญุณูุฉ
ps_sensors = []
failTrys = 0 
print("๐ ุฌุงุฑู ุงูุจุญุซ ุนู ุงูุญุณุงุณุงุช...")

# ูุญุงููุฉ ุงูุญุตูู ุนูู ุงูุญุณุงุณุงุช ุจุงูุฃุณูุงุก ุงูููุงุณูุฉ
for i in range(1,7):
    sensor_name = f'so{i}'
    sensor = robot.getDevice(sensor_name)
    if sensor:
        sensor.enable(timestep)
        ps_sensors.append(sensor)
        print(f"โ ุชู ุงูุนุซูุฑ ุนูู ุงูุญุณุงุณ ูุชูุนููู: {sensor_name}")
    else:
        # ุฅุฐุง ูู ูุฌุฏ 'so', ูุฏ ุชููู ุงูุฃุณูุงุก 'ps' ุฃู 'distance sensor'
        alt_name = f'ps{i}'
        sensor = robot.getDevice(alt_name)
        if sensor:
            sensor.enable(timestep)
            ps_sensors.append(sensor)
            print(f"โ ุชู ุงูุนุซูุฑ ุนูู ุญุณุงุณ ุจุฏูู: {alt_name}")

if not ps_sensors:
    print("โ ุฎุทุฃ ุญุฑุฌ: ูู ูุชู ุงูุนุซูุฑ ุนูู ุฃู ุญุณุงุณุงุช ูุณุงูุฉ! ุชุฃูุฏ ูู ุฃุณูุงุก ุงูุญุณุงุณุงุช ูู ุดุฌุฑุฉ ุงูุฑูุจูุช (Scene Tree).")
def set_speeds(l, r):
    left_motor.setVelocity(l)
    right_motor.setVelocity(r)

def get_color_name(c):
    if c[0] > 0.6 and c[1] < 0.4: return "RED"
    elif c[0] > 0.6 and c[1] > 0.6: return "YELLOW"
    elif c[1] > 0.6: return "GREEN"
    elif c[2] > 0.6: return "BLUE"
    return "UNKNOWN"



# ================== ูุชุบูุฑุงุช ุงูุชุญูู ==================
state = "SCAN_FOR_MISMATCH"
is_aerial_target = False  # ูุชุบูุฑ ุฌุฏูุฏ: ูู ุงููุฏู ูุฑููุน ูู ุงูููุงุกุ
previous_state = ""  # ูุชุบูุฑ ูุญูุธ ุงูุญุงูุฉ ูุจู ุงููุฑูุจ
pickup_color = ""        
delivery_color = ""      
# โ (ุฅุถุงูุฉ ุฌุฏูุฏุฉ) ูุตูููุฉ ูุญูุธ ุงูุฃููุงู ุงูุชู ุชู ููููุง
delivered_colors_list = []
timer = 0
has_reached_peak = False # (ูู ูุนุฏ ูุณุชุฎุฏููุง ูููู ุชุฑูุชูุง ููุชูุงูู)

# ูุณุงูุงุช ุงูุชููู
STOP_DISTANCE_CUBE = 0.075        
STOP_DISTANCE_TARGET = 0.15 

# ูุถุนูุงุช ุงูุฐุฑุงุน
ARM_UP_POS = -1.5    
ARM_DOWN_POS = 0.8   

# ุงูููุฉ
WALL_FORCE_LIMIT = 5.0 

print("๐ Pioneer 3-DX2: ุฌุงูุฒ (ุชุชุจุน ุฐูู + ุชูุงุฏู ุฑูุจูุชุงุช)...")
arm_pitch.setPosition(ARM_UP_POS)

while robot.step(timestep) != -1:
    try:
# ====================================================
        # ๐จ ูุธุงู ุงูุญูุงูุฉ ุงููุนุชูุฏ ุนูู ุงูููู (800+ = ุฎุทุฑ)
        # ====================================================
        obstacle_detected = False
        
        # ููุญุต ุงูุญุณุงุณุงุช ููุท ุฅุฐุง ูู ููู ูู ูุถุนูุฉ ุงูุฅูุณุงู ุฃู ุงูุฅููุงุช ุงูุฏูููุฉ
        if state not in ["PICKUP_ACTION", "DROP_ACTION", "AVOID_ROBOT" ,"AERIAL_PICKUP" ,"APPROACH_PARTNER" , "SEARCH_PARTNER", "HANDOVER_PREPARE"]:
            for i in range(6):
                val = ps_sensors[i].getValue()
                
                # ุจูุงุกู ุนูู ูุฑุงุกุงุชู: 1000 ูู ุงุตุทุฏุงูุ ูุฐุง 800 ูู ูุณุงูุฉ ุฃูุงู ูุงููุฉ
                threshold = 800.0 
                
                # ุงุณุชุซูุงุก: ุฅุฐุง ููุง ูุจุญุซ ุนู ููุนุจุ ูุชุฌุงูู ุงูุญุณุงุณุงุช ุงูุฃูุงููุฉ (3 ู 4) 
                # ุญุชู ูุง ููุฑุจ ูู ุงูููุนุจ ููุณู ุฅูุง ุฅุฐุง ุงูุชุตููุง ุจู (950)
                if state == "SEARCH_CUBE" and i in [3, 4]:
                    threshold = 950.0 
                
                if val > threshold:
                    obstacle_detected = True
                    # print(f"๐จ ุฎุทุฑ! ุงูุญุณุงุณ so{i} ูุฑุฃ {val:.2f}. ุชูุนูู ุงููุฑูุจ!")
                    break
        
        if obstacle_detected:
            previous_state = state
            state = "AVOID_ROBOT"
            timer = 0
            set_speeds(-1.0, -1.0) # ุชุฑุงุฌุน ููุฑู
            continue
# ====================================================
        # ๐ฎ ุขูุฉ ุงูุญุงูุงุช (State Machine) - ุงูุชุฑุชูุจ ุงูุตุญูุญ
        # ====================================================

        # 0) ุญุงูุฉ ุชูุงุฏู ุงูุฑูุจูุชุงุช (ูุฌุจ ุฃู ุชููู ุฃูู elif ุจุนุฏ ูุธุงู ุงูุญูุงูุฉ)
# 0) ุญุงูุฉ ุชูุงุฏู ุงูุฑูุจูุชุงุช
        if state == "AVOID_ROBOT":
            timer += 1
            if timer < 40:
                set_speeds(-1.2, -1.2) # ุชุฑุงุฌุน ุณุฑูุน ููุฎูู
            elif timer < 80:
                set_speeds(1.0, -1.0)  # ุฏูุฑุงู ุญุงุฏ ููุงุจุชุนุงุฏ ุนู ุงูุนุงุฆู
            else:
                print(f"โ ุชู ุงูุงุจุชุนุงุฏุ ุงูุนูุฏุฉ ููุญุงูุฉ: {previous_state}")
                state = previous_state
                timer = 0
                set_speeds(0, 0)
        # 1) ุงูุจุญุซ ุงูุนุงู
# 1) ุงูุจุญุซ ุงูุนุงู (ุชู ุชุนุฏูู ููุทู ุชุญุฏูุฏ ุงููุฏู)
# ================================================================
# ุชุนุฏูู: ุฏุงุฎู SCAN_FOR_MISMATCH ููุชุญููู ุฅูู ุงูุจุญุซ ุนู ุงูุฑูุจูุช
# ================================================================
        elif state == "SCAN_FOR_MISMATCH":
            is_aerial_target = False  # ุงูููุนุจุงุช ููุง ุฃุฑุถูุฉ
            arm_pitch.setPosition(ARM_UP_POS)
            
            # โญ ุงูุดุฑุท ุงูุฌุฏูุฏ: ุฅุฐุง ุณูููุง 3 ููุนุจุงุชุ ูุจุญุซ ุนู ุงูุฑูุจูุช ุจุฏูุงู ูู ุงูููุนุจุงุช
            if delivered_cubes >= 3:
                print("๐ค ุงูุชูุช ุงูููุนุจุงุช ุงูุฃุฑุถูุฉ.. ุงูุจุญุซ ุนู ุงูุฑูุจูุช ุงูุดุฑูู!")
                state = "SEARCH_PARTNER" # ุญุงูุฉ ุฌุฏูุฏุฉ ุณูุถูููุง
                wrist_roll.setPosition(0.0) # ุงููุนุตู ุฃููู ูููุชูุท ุงูููุนุจ ุงูุนููุฏู
                timer = 0
                continue # ุชุฎุทู ุจููุฉ ุงูููุฏ ูุงูุงูุชูุงู ููุญุงูุฉ ุงูุชุงููุฉ

            # --- (ุจููุฉ ููุฏ ุงูุจุญุซ ุนู ุงูููุนุจุงุช ููุง ูู ุฏูู ุชุบููุฑ) ---
            set_speeds(0.5, -0.5) 
            objs = cam_reg.getRecognitionObjects()
            cubes = [o for o in objs if o.getModel() != "TARGET"]
            targets = [o for o in objs if o.getModel() == "TARGET"]

            for cube in cubes:
                for target in targets:
                    c_pos = cube.getPosition()
                    t_pos = target.getPosition()
                    width_diff = abs(c_pos[1] - t_pos[1])
                    c_depth = abs(c_pos[2])
                    t_depth = abs(t_pos[2])
                    depth_gap = t_depth - c_depth

                    if 0 < depth_gap < 0.3 and width_diff < 0.08:
                        c_color = get_color_name(cube.getColors())
                        t_color = get_color_name(target.getColors())
                        
                        if c_color != t_color: 
                            pickup_color = c_color
                            delivery_color = c_color 
                            print(f"๐ฏ ุชู ุฑุตุฏ ุชุทุงุจู: {c_color}")
                            state = "SEARCH_CUBE"
                            break
        # 2) ุงูุชูุฌู ููููุนุจ (ุชุชุจุน ุฏููู)
        elif state == "SEARCH_CUBE":
            arm_pitch.setPosition(ARM_UP_POS)
            gripper_left.setPosition(0.09) # ุชุฃูุฏ ุฃู ุงููููุท ููุชูุญ
            gripper_right.setPosition(0.11)

            objs = cam_reg.getRecognitionObjects()
            target_obj = None
            if objs:
                for o in objs:
                    if o.getModel() != "TARGET" and get_color_name(o.getColors()) == pickup_color:
                        target_obj = o
                        break
            
            if target_obj:
                pos = target_obj.getPosition()
                side_deviation = pos[1] 
                distance_to_obj = abs(pos[2])

                # ููุทู ุงููุญุงุฐุงุฉ (ุชุตููุฑ ุงูุงูุญุฑุงู)
                if abs(side_deviation) > 0.04: 
                    if side_deviation > 0:
                        set_speeds(-0.2, 0.2)
                    else:
                        set_speeds(0.2, -0.2)
                else:
                    if is_aerial_target:
                        # ุงูุชูุฏู
                        if distance_to_obj > STOP_DISTANCE_CUBE+0.05:
                            set_speeds(0.7, 0.7)
                        else:
                            set_speeds(0, 0)
                            state = "PICKUP_ACTION"
                            timer = 0
                    else: 
                        if distance_to_obj > STOP_DISTANCE_CUBE:
                            set_speeds(0.7, 0.7)
                        else:
                            set_speeds(0, 0)
                            state = "PICKUP_ACTION"
                            timer = 0
                        
            else:
                set_speeds(0.4, -0.4) 
# 3) ุงูุงูุชูุงุท
        elif state == "PICKUP_ACTION":
            timer += 1
            if timer == 1:
                print("๐ ูุชุญ ุงููููุท")
                gripper_left.setPosition(0.11) 
                gripper_right.setPosition(0.11)
                
            elif timer == 40:
                if is_aerial_target:
                    print("โ๏ธ ุชูุฌูู ุงูุฐุฑุงุน ููุฃูุงู (ุงูุชูุงุท ููุงุฆู)")
                    arm_pitch.setPosition(0.6)  # 0.6 ุชูุฑูุจุงู ูุณุชูู ุฃููู (ุนุฏููุง ุญุณุจ ุงุฑุชูุงุน ุดุฑููู)
                else:
                    print("๐ฝ ุฅูุฒุงู ุงูุฐุฑุงุน ููุฃุฑุถ (ุงูุชูุงุท ุฃุฑุถู)")
                    arm_pitch.setPosition(ARM_DOWN_POS) # 0.8 ููุฃุฑุถ
            elif timer == 100: 
                print("โ ุฅุบูุงู ุงููููุท")
                gripper_left.setPosition(0.0) 
                gripper_right.setPosition(0.0) 
            
            elif timer == 160:
                # ูุญุต ุงูุญุณุงุณุงุช
                left_f = gl_sensor.getValue()
                right_f = gr_sensor.getValue()
                max_force = max(left_f, right_f)

                if max_force > WALL_FORCE_LIMIT: 
                    print(f"๐จ ุฌุฏุงุฑ! ุฅููุงุช...")
                    gripper_left.setPosition(0.11)
                    gripper_right.setPosition(0.11)
                    state = "RETRY_MOVE"
                    timer = 0
                elif left_f > 0.002 or right_f > 0.002:
                    print(f"๐ฆ ูุฌุงุญ ุงูุฅูุณุงู ุงููุจุฏุฆู ({max_force:.3f})... ุฌุงุฑู ุงูุฑูุน ูุงูุชุญูู")
                    failTrys = 0
                    state = "LIFTING"
                    timer = 0
                else:
                    print("โ๏ธ ูุดู (ูุฑุงุบ)...")
                    failTrys += 1
                    state = "RETRY_MOVE"
                    timer = 0
                    if failTrys > 3:
                        print("๐งญ ูุญุงููุงุช ูุงุดูุฉ โ ุฌุนู ุงููููุท ุฃููู")
                        wrist_roll.setPosition(0.0) 

        # 4) ุงูุฑูุน ูุน ุงูุชุญูู ุงููุณุชูุฑ
        elif state == "LIFTING":
            timer += 1
            # ุฅุฌุจุงุฑ ุงููููุท ุนูู ุงูุจูุงุก ูุบููุงู
            gripper_left.setPosition(0.0)
            gripper_right.setPosition(0.0)
            
            # ุฑูุน ุงูุฐุฑุงุน
            arm_pitch.setVelocity(0.8)
            arm_pitch.setPosition(ARM_UP_POS) 

            # --- ุงูุชุนุฏูู ุงูุฌุฏูุฏ: ูุญุต ุงูุฃูุงู ุฃุซูุงุก ุงูุฑูุน ---
            if timer > 20: # ููุชุธุฑ ููููุงู ุจุนุฏ ุจุฏุก ุงูุฑูุน ููุชุฃูุฏ ูู ุงุณุชูุฑุงุฑ ุงููุฑุงุกุฉ
                left_f = gl_sensor.getValue()
                right_f = gr_sensor.getValue()
                
                # ุฅุฐุง ููุฏูุง ุงูุถุบุท ุชูุงูุงู ุฃุซูุงุก ุงูุฑูุน
                if left_f < 0.001 and right_f < 0.001:
                    print("โ ููุฏุงู ุงูููุนุจ ุฃุซูุงุก ุงูุฑูุน! ุฅุนุงุฏุฉ ุงููุญุงููุฉ...")
                    state = "RETRY_MOVE" # ุฃู ุงุฑุฌุน ูุญุงูุฉ SEARCH_CUBE ูุจุงุดุฑุฉ
                    timer = 0
                    # ููุชุญ ุงููููุท ูุชุฌูุจ ุงูุชุนููู
                    gripper_left.setPosition(0.11)
                    gripper_right.setPosition(0.11)

            # ุงูุงูุชูุงุฑ ุญุชู ุชุฑุชูุน ุงูุฐุฑุงุน ุชูุงูุงู
            if timer > 100: 
                print(f"โ ุชู ุงูุชุฃูุฏ ูู ูุฌูุฏ ุงูููุนุจ. ุงูุจุญุซ ุนู ุงููุฏู {delivery_color}...")
                state = "SEARCH_TARGET"
                timer = 0
        # 5) ุงูุจุญุซ ุนู ุงููุฏู (ุชู ุชุญุฏูุซู ููุนูู ูุซู ุงูุจุญุซ ุนู ุงูููุนุจ)
        elif state == "SEARCH_TARGET":
            # ุญุงูุธ ุนูู ูุถุนูุฉ ุงูุญูู
            arm_pitch.setPosition(ARM_UP_POS)
            gripper_left.setPosition(0.0)
            gripper_right.setPosition(0.0)

            objs = cam_reg.getRecognitionObjects()
            target_obj = None
            if objs:
                for o in objs:
                    # ูุจุญุซ ุนู ูุงุฆู ููุนู TARGET ููู ุงูููู ุงููุทููุจ
                    if o.getModel() == "TARGET" and get_color_name(o.getColors()) == delivery_color:
                        target_obj = o
                        break
            
            if target_obj:
                pos = target_obj.getPosition()
                side_deviation = pos[1] # ุงูุงูุญุฑุงู ุงูุฌุงูุจู
                distance_to_obj = abs(pos[2])

                # ููุณ ููุทู ุงููุญุงุฐุงุฉ ุงูุฏููู ุงููุณุชุฎุฏู ูุน ุงูููุนุจ
                if abs(side_deviation) > 0.05: # ูุงูุด ุฃูุจุฑ ููููุงู ูููุฏู
                    if side_deviation > 0:
                        set_speeds(-0.25, 0.25)
                    else:
                        set_speeds(0.25, -0.25)
                else:
                    if distance_to_obj > STOP_DISTANCE_TARGET:
                        set_speeds(0.7, 0.7)
                    else:
                        set_speeds(0, 0)
                        state = "DROP_ACTION"
                        timer = 0
            else:
                set_speeds(0.4, -0.4) # ุฏูุฑุงู ููุจุญุซ

        # 6) ูุถุน ุงูููุนุจ
        elif state == "DROP_ACTION":
            timer += 1
            if timer == 20:
                print("๐ฝ ุฅูุฒุงู ููุถุน ุงูููุนุจ...")
                arm_pitch.setPosition(ARM_DOWN_POS)
            elif timer == 70: 
                print("๐ ุฅููุงุช...")
                gripper_left.setPosition(0.09)
                gripper_right.setPosition(0.09)
            elif timer == 100:
                print("๐ผ ุงุจุชุนุงุฏ...")
                arm_pitch.setPosition(ARM_UP_POS)
                set_speeds(-0.8, -0.8)
            elif timer == 150:
                set_speeds(0, 0)
                delivered_cubes += 1
                
                # โ (ุฅุถุงูุฉ ุฌุฏูุฏุฉ) ุชุณุฌูู ููู ุงูููุนุจ ุงูุฐู ุชู ุชุณูููู ูู ุงููุงุฆูุฉ
                delivered_colors_list.append(delivery_color)
                print(f"๐ ุชู ุชุณุฌูู ุงูููู {delivery_color}. ุงููุงุฆูุฉ: {delivered_colors_list}")
                
                print(f"๐ฆ ุชู ุชุณููู {delivered_cubes} ููุนุจ/ููุนุจุงุช")
            
                # โญ ุงูุดุฑุท ุงููุทููุจ
                if delivered_cubes > 3:
                    print("๐งญ ุฅุชูุงู ุงููููุฉ โ ุฌุนู ุงููููุท ุนููุฏู")
                    wrist_roll.setPosition(1.57)  # 90 ุฏุฑุฌุฉ
                else:
                    wrist_roll.setPosition(0.0)   # ูุจูู ุฃููููุง
            
                state = "SCAN_FOR_MISMATCH"
                timer = 0

        # 7) ุงููุญุงููุฉ ูุฑุฉ ุฃุฎุฑู (ุนูุฏ ุงููุดู ุฃู ุงูุฌุฏุงุฑ)
        elif state == "RETRY_MOVE":
            timer += 1
            arm_pitch.setPosition(ARM_UP_POS)
            if timer < 50:
                set_speeds(-0.8, -0.8) # ุชุฑุงุฌุน
            elif timer < 90:
                set_speeds(0.6, -0.6) # ุฏูุฑุงู ูุชุบููุฑ ุงูุฒุงููุฉ
            else:
                state = "SEARCH_CUBE" # ุฅุนุงุฏุฉ ุงูุจุญุซ
                timer = 0
                
                
                # ================================================================
# ุฅุถุงูุฉ: ุญุงูุงุช ุงูุชูุงุท ุงูููุนุจ ูู ุงูุฑูุจูุช (Handover States)
# ================================================================
        
        # 1) ุงูุจุญุซ ุนู ุงูุฑูุจูุช ุงูุฃูู
        elif state == "SEARCH_PARTNER":
            set_speeds(0.4, -0.4) # ุฏูุฑุงู ููุจุญุซ
            objs = top_cam.getRecognitionObjects()
            partner = None
            if objs:
                for o in objs:
                    if o.getModel() == "ROBOT": # ุงูุจุญุซ ุนู ุงูุฑูุจูุช
                        partner = o
                        break
            
            if partner:
                print(" ุชู ุฑุตุฏ ุงูุดุฑูู.. ")
    # ูุณุชุฎุฏู ุงููุงููุฑุง ุงูุฃูุงููุฉ cam_reg ูุถูุงู ุฃููุง ูุฑู ุงูููุนุจ ููุง ูุฑู ุงูููุนุจ ุงูุฃุฑุถู
                objs = cam_reg.getRecognitionObjects()
                found_cube = None
                
                # ูุจุญุซ ุนู ุฃู ููุนุจ (ููุณ TARGET)
                for o in objs:
                    if o.getModel() != "TARGET" and o.getModel() != "ROBOT":
    # ุงูุญุตูู ุนูู ููู ุงูููุนุจ ุงูุฐู ูุญููู ุงูุดุฑูู
                        detected_color = get_color_name(o.getColors())
                        
                        # โ๏ธ ุงูุดุฑุท: ูู ูุฐุง ุงูููู ููุฌูุฏ ูู ุงููุงุฆูุฉุ
                        if detected_color in delivered_colors_list:
                            continue # ุชุฎุทู ูุฐุง ุงูููุนุจ ูุฃูู ูููู ุณุงุจูุงู
                        else:
                            found_cube = o # ูุฌุฏูุง ููุนุจุงู ุฌุฏูุฏุงู
                            break                
                if found_cube:
                    c_color = get_color_name(found_cube.getColors())
                    print(f"๐ ุชู ุฑุตุฏ ููุนุจ ูุน ุงูุดุฑูู ุจููู {c_color}.. ุชูุนูู ุงูุชุชุจุน ุงูุนุงุฏู!")
                    
                    pickup_color = c_color      # ุชุญุฏูุฏ ุงูููู ููุงูุชูุงุท
                    delivery_color = c_color    # ุชุญุฏูุฏ ููู ุงููุฏู ูุงุญูุงู
                    
                    is_aerial_target = True     # โ๏ธ ูุงู ุฌุฏุงู: ุฃุฎุจุฑูุง ุงูุฑูุจูุช ุฃู ูุฐุง ุงููุฏู ูุฑุชูุน
                    
                    state = "SEARCH_CUBE"       # โ ููุชูู ูููุณ ุญุงูุฉ ุชุชุจุน ุงูููุนุจ ุงูุฃุฑุถู
                    timer = 0
                    set_speeds(0, 0)
# 2) ุงูุงูุชุฑุงุจ ุงูุฏููู ูู ุงูุฑูุจูุช
# 2) ุงูุงูุชุฑุงุจ ุงูุฏููู ูู ุงูุฑูุจูุช (ุฏูุฌ ุงููุงููุฑุง ูุน ุงูุญุณุงุณุงุช)
        elif state == "APPROACH_PARTNER":
            # 1. ุชุฃูุฏ ูู ุฑูุน ุงูุฐุฑุงุน ุญุชู ูุง ุชุดูุด ุนูู ุงููุงููุฑุง ุฃู ุงูุญุณุงุณุงุช
            arm_pitch.setPosition(ARM_UP_POS) 
            
            # 2. ูุฑุงุกุฉ ููู ุงูุญุณุงุณุงุช ุงูุฃูุงููุฉ (so3 ู so4 ููุง ุงูุฃูุณุท ูู Pioneer)
            # ููุงุญุธุฉ: ูู ููุฏู ุงูุณุงุจูุ ุงููููุฉ ุงูุนุงููุฉ ุชุนูู ุนุงุฆู ูุฑูุจ ( > 800 ุฎุทุฑ)
            # ุงูุญุณุงุณุงุช: ps_sensors[3] ู ps_sensors[4]
            front_val_left = ps_sensors[3].getValue()
            front_val_right = ps_sensors[4].getValue()
            
            # ูุฃุฎุฐ ุงููุชูุณุท ุฃู ุงูุฃูุจุฑ ุจููููุง ูุถูุงู ุฃููุง ูุฑู ุงูุฌุณู
            avg_dist_sensor = (front_val_left + front_val_right) / 2
            
            # 3. ุงุณุชุฎุฏุงู ุงููุงููุฑุง ููุชูุฌูู ููุท (ูููู/ูุณุงุฑ)
            objs = top_cam.getRecognitionObjects()
            partner = None
            
            if objs:
                for o in objs:
                    if o.getModel() == "ROBOT": 
                        partner = o
                        break
            
            if partner:
                pos = partner.getPosition()
                side = pos[1] # ุงูุงูุญุฑุงู ุงูุฌุงูุจู (ุญุณุจ ุชุฌุฑุจุชู ุงููุงุฌุญุฉ)
                
                # --- [ุฃ] ููุทู ุงูุชูุฌูู (ุงููุงููุฑุง) ---
                if abs(side) > 0.04:
                    if side > 0: set_speeds(-0.2, 0.2)
                    else: set_speeds(0.2, -0.2)
                    
                # --- [ุจ] ููุทู ุงูุชูุฏู (ุงูุญุณุงุณุงุช) ---
                else:
                    # ุฃูุช ููุช ุณุงุจูุงู ุฃู 800 ุชุนูู ุงุตุทุฏุงู/ุฎุทุฑ
                    # ูุฑูุฏ ุงูุชููู ูุจู ุงูุงุตุทุฏุงู ุจูููู ูููู ูุฑูุจูู ุฌุฏุงู ูููุตุงูุญุฉ
                    # ุฌุฑุจ ูููุฉ ุจูู 300 ุฅูู 500 (ูููุง ุฒุงุฏุช ุงููููุฉุ ุงูุชุฑุจ ุงูุฑูุจูุช ุฃูุซุฑ)
                    STOP_SENSOR_THRESHOLD = 400.0 
                    
                    print(f"๐ ุงูุญุณุงุณุงุช: {avg_dist_sensor:.1f} | ุงููุงููุฑุง Side: {side:.3f}")

                    if avg_dist_sensor < STOP_SENSOR_THRESHOLD:
                        # ูู ูุตู ุจุนุฏ (ุงููููุฉ ููุฎูุถุฉ = ุงูุทุฑูู ูุงุฑุบ ุฃูุงููุง)
                        set_speeds(0.5, 0.5)
                    else:
                        # ุงููููุฉ ุชุฌุงูุฒุช ุงูุญุฏ = ูุญู ูุฑูุจูู ุฌุฏุงู ูู ุงูุฑูุจูุช ุงูุขุฎุฑ
                        set_speeds(0, 0)
                        print(f"โ ุชููู ุจุงูุญุณุงุณุงุช ({avg_dist_sensor:.1f}).. ุงูุชุฌููุฒ ููุงุณุชูุงู")
                        state = "HANDOVER_PREPARE"
                        timer = 0
            
            else:
                # ุฅุฐุง ููุฏูุง ุงููุงููุฑุง ูููู ุงูุญุณุงุณุงุช ุชูุฑุฃ ุดูุฆุงู ูุฑูุจุงู ุฌุฏุงูุ ูุฏ ูููู ูุตููุง ุจุงููุนู!
                if avg_dist_sensor > 600:
                    print("โ๏ธ ุงููุงููุฑุง ูุง ุชุฑูุ ููู ุงูุญุณุงุณุงุช ุชููู ุฃููุง ูุตููุง!")
                    set_speeds(0, 0)
                    state = "HANDOVER_PREPARE"
                    timer = 0
                else:
                    print("๐ ุจุญุซ ุนู ุงูุดุฑูู...")
                    set_speeds(0.2, -0.2) # ุฏูุฑุงู ุจุทูุก ููุจุญุซ


        elif state == "HANDOVER_PREPARE":
            timer += 1
            set_speeds(0, 0)  # ุชุซุจูุช ุงูุชููู
        
            if timer == 1:
                print("๐ ุงูุขู ุฑูุน ุงูุฐุฑุงุน ููุถุนูุฉ ุงูุงุณุชูุงู (0.6)")
                # โ ุงูุขู ููุท ูุถุน ุงูุฐุฑุงุน ูู ุงููุถุน ุงูุฃููู ูุฃููุง ุชููููุง ููุตููุง
                arm_pitch.setPosition(0.6)  
                
                # ูุชุญ ุงููููุท
                gripper_left.setPosition(0.09)
                gripper_right.setPosition(0.09)
        
            # ููุชุธุฑ ููุชุงู ูุงููุงู (HANDOVER_ARM_DELAY = 25) ูุชุณุชูุฑ ุงูุฐุฑุงุน
            if timer > HANDOVER_ARM_DELAY:
                print("๐ค ุงูุฐุฑุงุน ุงุณุชูุฑุช.. ุงูุงูุชูุงู ููุงูุชูุงุท")
                state = "AERIAL_PICKUP"
                timer = 0
        # 3) ุชูููุฐ ุงูุงูุชูุงุท ุงูุฌูู
        elif state == "AERIAL_PICKUP":
            set_speeds(0, 0)
            timer += 1
            
            # ุชุซุจูุช ุงูุฐุฑุงุน ุฃูููุงู
            arm_pitch.setPosition(0.6)
            
            if timer == 20:
                print("โ ุฅุบูุงู ุงููููุท ุนูู ููุนุจ ุงูุฒููู")
                gripper_left.setPosition(0.0)
                gripper_right.setPosition(0.0)
            
            elif timer == 80:
                # ุงูุชุญูู ูู ุฃูุณููุง ุดูุฆุงูุ
                l_val = gl_sensor.getValue()
                r_val = gr_sensor.getValue()
                if l_val > 0.002 or r_val > 0.002:
                    print(f"๐ฆ ูุฌุงุญ ุงูุงุณุชูุงู! ({max(l_val, r_val):.3f})")
                    # ููุชุฑุถ ุฃู ุงูููู ูู ุงูุฃุญูุฑ ููููุนุจ ุงูุฃุฎูุฑ ุฃู ุญุณุจ ุงูููุทู
                    delivery_color = "RED"  # ุฃู ุงุฌุนููุง ูุชุบูุฑ ุนุงู ุฅุฐุง ููุช ุชุนุฑู ุงูููู
                    state = "REVERSE_AND_DELIVER"
                else:
                    print("โ๏ธ ูุดู ุงูุงุณุชูุงู (ูุฑุงุบ).. ูุญุงููุฉ ุชูุฏู ุจุณูุท")
                    set_speeds(0.2, 0.2) # ุชูุฏู ููููุงู
                    if timer > 100: # ูููุฉ ููุงุฆูุฉ
                        timer = 0 # ุฅุนุงุฏุฉ ุฏูุฑุฉ ุงูุฅุบูุงู
            
        # 4) ุงูุฑุฌูุน ููุฎูู ูุงูุฐูุงุจ ููุชุณููู
        elif state == "REVERSE_AND_DELIVER":
            timer += 1
            # ุฑูุน ุงูุฐุฑุงุน ููุฃุนูู ูุญูุงูุฉ ุงูููุนุจ
            arm_pitch.setPosition(ARM_UP_POS)
            
            if timer < 40:
                set_speeds(-0.8, -0.8) # ุฑุฌูุน ููุฎูู ููุงููุตุงู ุนู ุงูุฑูุจูุช ุงูุฃูู
            else:
                print("๐ ุงูุชูุฌู ูููุฏู...")
                state = "SEARCH_TARGET" # ุงูุงูุชูุงู ูููุทู ุงูุชุณููู ุงูููุฌูุฏ ูุณุจูุงู
                timer = 0

    except Exception as e:
        print(f"โ๏ธ Error: {e}")
        break